pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -212,20 +245,220 @@ func (c *MCDefaultRouteController) processNextWorkItem() bool {
 	}
 	defer c.queue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
+	key, ok := obj.(string)
+	if !ok {
 		c.queue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+	if c.wireGuardConfig != nil {
+		gateway, err := c.getActiveGateway()
+		if err != nil {
+			klog.ErrorS(err, "Failed to get active Gateway")
+			c.queue.AddRateLimited(key)
+			return true
+		}
+
+		isMyselfGateway := gateway != nil && gateway.Name == c.nodeConfig.Name
+		// If current Node is Multi-cluster Gateway. controller will initialize WireGuard configuration for Gateway event
+		// and add WireGuard peers for ClusterInfoImport event.
+		// If current Node is not Multi-cluster Gateway, then controller will not reconcile WireGuard.
+		// If it's the former Gateway, controller will try to remove corresponding configurations.
+		if isMyselfGateway {
+			isGatewayItem := isGatewayKey(key)
+			if !isGatewayItem && !c.wireGuardInitialized {
+				// If the WireGuard has not been initialized. Some necessary configuration may not
+				// be ready to add a peer. In this situation, we need to requeue the ClusterInfoImport again.
+				klog.V(2).InfoS("WireGuard configuration has not been reconciled, requeue")
+				c.queue.AddRateLimited(key)
+				return true
+			}
+			wireGuardWasInitialized := c.wireGuardInitialized
+			if err := c.reconcileWireGuard(key, isGatewayItem, isMyselfGateway); err != nil {
+				klog.ErrorS(err, "Failed to reconcile Multi-cluster WireGuard")
+				c.queue.AddRateLimited(key)
+				return true
+			}
+
```

Comment:
> Yes, the Gateway will be updated with a patch and re-queued immediately. So I think maybe it's not necessary to syncFlows.
> 

But what is the harm to add flows earlier? And again, later if we check Gateway attribute changes, we can ignore events for a public key change. I am ok if you still want to keep the current way, but as it is hard to figure out the reasoning, at least make it clear in the comments that the Gateway update is caused by adding the WG key to the Gateway resource.

> As for the real change check, it's a good way to solve this. Maybe we can compare the old obj and new obj to determine enqueue or not.

Yes, and we can do further to check what attributes changed and decide what operations to do based on that. BTW, I do not mean we must do these in this PR.

Reply:
Yes this block can be removed. If we do not care add flows earlier, the `wireGuardWasInitialized` check is removed now.